### PR TITLE
stm32: add support for backup SRAM

### PR DESCRIPTION
**Pull request by gmarull**
Add support for backup SRAM initialization found in multiple STM32
microcontrollers. Linker script facilities are also provided to make it
easy to define variables in the backup SRAM.

### CODE REVIEW

File: samples/memc/stm32_backup_sram/src/main.c [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+{
+	printk("Current value in backup SRAM: %d\n", backup_value);
+
+	backup_value++;
+
+	printk("Next reported value should be: %d\n", backup_value);
+	printk("Re-power the board now (excepting VBAT)\n");
+
+#if __DCACHE_PRESENT
+	SCB_CleanDCache_by_Addr(&backup_value, sizeof(backup_value));
```
Comment (gmarull) [inline samples/memc/stm32_backup_sram/src/main.c:23]: Seems to be required on M7 (see https://community.st.com/s/question/0D50X0000ALwOvkSQF/i-can-not-use-backup-ram-for-stm32h743bi-is-there-any-application-note-or-source-code-for-backup-ram). Not sure if there is a better/more generic solution.

File: samples/memc/stm32_backup_sram/README.rst [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+NVM when VBAT pin is supplied with a voltage source, e.g. a coin button cell.
+
+This example shows how to define a variable on the Backup SRAM. Each time the
+application runs the current value is displayed and then incremented by one. If
+VBAT is preserved the incremented value will be shown on the next power-cycle.
+
+.. note::
+
+    On some boards VBAT is connected to VDD. You may want to remove the
+    connection jumper if using an independent VBAT source.
```
Comment (erwango) [inline samples/memc/stm32_backup_sram/README.rst:19]: This part could be clearer.
```suggestion
    On most boards, VBAT is typically connected to VDD thanks to a jumper.
    To exercise this sample with independent VBAT source, you will need to remove the jumper.
```

File: samples/memc/stm32_backup_sram/src/main.c [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+__stm32_backup_sram_section uint32_t backup_value;
+
+void main(void)
+{
+	printk("Current value in backup SRAM: %d\n", backup_value);
+
+	backup_value++;
+
+	printk("Next reported value should be: %d\n", backup_value);
+	printk("Re-power the board now (excepting VBAT)\n");
```
Comment (erwango) [inline samples/memc/stm32_backup_sram/src/main.c:20]: This could be misleading. What about `Reset the board` ?

File: drivers/memc/Kconfig.stm32 [hunk 4]
```diff
@@ -17,3 +17,10 @@ config MEMC_STM32_SDRAM
 	select USE_STM32_HAL_SDRAM
 	help
 	  Enable STM32 FMC SDRAM controller.
+
+config MEMC_STM32_BACKUP_SRAM
+	bool "Enable STM32 Backup SRAM"
+	depends on SOC_SERIES_STM32H7X
```
Comment (erwango) [inline drivers/memc/Kconfig.stm32:23]: I'd prefer we link this directly with device tree thanks to:
```suggestion
        bool
	default default $(dt_compat_enabled,$(DT_COMPAT_ST_STM32_BACKUP_SRAM))
```
And we get rid of `MEMC_STM32_BACKUP_SRAM` symbol selection by the user.

File: samples/memc/stm32_backup_sram/README.rst [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+Overview
+********
+
+Multiple STM32 microcontrollers have a small backup SRAM that can be used as a
+NVM when VBAT pin is supplied with a voltage source, e.g. a coin button cell.
+
+This example shows how to define a variable on the Backup SRAM. Each time the
+application runs the current value is displayed and then incremented by one. If
+VBAT is preserved the incremented value will be shown on the next power-cycle.
```
Comment (erwango) [inline samples/memc/stm32_backup_sram/README.rst:14]:
```suggestion
application runs the current value is displayed and then incremented by one. If
VBAT is preserved, the incremented value will be shown on the next power-cycle.
```

**Review by erwango — CHANGES_REQUESTED** [review]
Few comments, but otherwise a valuable addition, thanks!

File: drivers/memc/Kconfig.stm32 (hunk 4 above)
Comment (gmarull) [reply drivers/memc/Kconfig.stm32:23]: If we do this, then all boards with

```
&backup_sram {
    status = "okay";
};
```

will have the backup SRAM enabled. I'm not sure if this is what we want.

File: samples/memc/stm32_backup_sram/README.rst (hunk 2 above)
Comment (gmarull) [reply samples/memc/stm32_backup_sram/README.rst]: fixed, thanks

File: samples/memc/stm32_backup_sram/src/main.c (hunk 3 above)
Comment (gmarull) [reply samples/memc/stm32_backup_sram/src/main.c]: Replaced with `Unplug the board`. Reset may indicate pressing the reset button, which is a valid test but does not cover retention with VBAT only.

File: samples/memc/stm32_backup_sram/README.rst (hunk 5 above)
Comment (gmarull) [reply samples/memc/stm32_backup_sram/README.rst]: fixed, thanks

File: drivers/memc/Kconfig.stm32 (hunk 4 above)
Comment (erwango) [reply drivers/memc/Kconfig.stm32:23]: Right, application control is still required.

File: samples/memc/stm32_backup_sram/src/main.c (hunk 3 above)
Comment (erwango) [reply samples/memc/stm32_backup_sram/src/main.c]: On most ST boards, `Unplug the board` would unpower VBAT too, right ?

Comment (gmarull) [reply samples/memc/stm32_backup_sram/src/main.c]: you're right... in fact on these boards the VBAT mode can't be tested...

Comment (gmarull) [reply samples/memc/stm32_backup_sram/src/main.c]: what about now: `Keep VBAT power source and reset the board now!`?

**Comment by gmarull** [comment]
@erwango because `MEMC_STM32` is auto-selected for `MEMC=y` and STM32, it makes me think if `memc` is the right place for this "driver" or it should be part of `soc` (now FMC gets enabled...). If it's kept under `memc`, we can't perform the auto-selection, and we should probably rename the driver to `MEMC_STM32_FMC` to make a clear distinction.

File: samples/memc/stm32_backup_sram/src/main.c (hunk 3 above)
Comment (erwango) [reply samples/memc/stm32_backup_sram/src/main.c]: Sounds good!

**Comment by erwango** [comment]
> @erwango because `MEMC_STM32` is auto-selected for `MEMC=y` and STM32, it makes me think if `memc` is the right place for this "driver" or it should be part of `soc` (now FMC gets enabled...). If it's kept under `memc`, we can't perform the auto-selection, and we should probably rename the driver to `MEMC_STM32_FMC` to make a clear distinction.

@gmarull I agree there is no reason to mix FMC and backup RAM. I'm fine with putting this in `soc`. This means the sample will have to be moved as well (to eg `samples/boards/stm32`).

**Comment by gmarull** [comment]
@erwango @ABOSTM ping

**Review by ABOSTM — APPROVED** [review]
LGTM

File: samples/boards/stm32/backup_sram/sample.yaml [hunk 6]
```diff
@@ -0,0 +1,8 @@
+sample:
+  description: Usage of backup SRAM on STM32
+  name: stm32_backup_sram
+tests:
+  sample.boards.stm32.backup_sram:
+    depends_on: backup_sram
+    build_only: true
```
Comment (erwango) [inline samples/boards/stm32/backup_sram/sample.yaml:7]: Would be nice to make it runnable in CI, but this is next step.

**Review by erwango — APPROVED** [review]
Thanks for this addition!

File: samples/boards/stm32/backup_sram/sample.yaml (hunk 6 above)
Comment (gmarull) [reply samples/boards/stm32/backup_sram/sample.yaml:7]: I guess we need hardware on CI for that?

Comment (erwango) [reply samples/boards/stm32/backup_sram/sample.yaml:7]: We have a bench and daily tests are run.

**Pull request merged**